pallets/presto/src/requests.rs

```diff
@@ -0,0 +1,260 @@
+use codec::{Decode, Encode, MaxEncodedLen};
+use common::{AccountIdOf, Balance, BoundedString};
+use frame_support::traits::Time;
+use sp_core::RuntimeDebug;
+use sp_runtime::{DispatchError, DispatchResult};
+
+#[derive(
+    RuntimeDebug, Clone, PartialEq, Eq, Encode, Decode, scale_info::TypeInfo, MaxEncodedLen,
+)]
+#[scale_info(skip_type_params(T))]
+pub enum RequestStatus<T: Config> {
+    Pending,
+    Cancelled,
+    Approved {
+        by: AccountIdOf<T>,
+        time: MomentOf<T>,
+    },
+    Declined {
+        by: AccountIdOf<T>,
+        time: MomentOf<T>,
+    },
+}
+
+#[derive(
+    RuntimeDebug, Clone, PartialEq, Eq, Encode, Decode, scale_info::TypeInfo, MaxEncodedLen,
+)]
+#[scale_info(skip_type_params(T))]
+pub enum Request<T: Config> {
+    Deposit(DepositRequest<T>),
+    Withdraw(WithdrawRequest<T>),
+}
+
+impl<T: Config> Request<T> {
+    pub fn owner(&self) -> &AccountIdOf<T> {
+        match self {
+            Self::Deposit(request) => &request.owner,
+            Self::Withdraw(request) => &request.owner,
+        }
+    }
+
+    pub fn status(&self) -> &RequestStatus<T> {
+        match self {
+            Self::Deposit(request) => &request.status,
+            Self::Withdraw(request) => &request.status,
+        }
+    }
+
+    pub fn decline(&mut self, manager: AccountIdOf<T>) -> DispatchResult {
+        match self {
+            Self::Deposit(request) => request.decline(manager),
+            Self::Withdraw(request) => request.decline(manager)?,
+        }
+
+        Ok(())
+    }
+
+    pub fn cancel(&mut self) -> DispatchResult {
+        match self {
+            Self::Deposit(request) => request.cancel(),
+            Self::Withdraw(request) => request.cancel()?,
+        }
+
+        Ok(())
+    }
+}
+
+#[derive(RuntimeDebug, Encode, Decode, scale_info::TypeInfo, MaxEncodedLen)]
+#[scale_info(skip_type_params(T))]
+pub struct DepositRequest<T: Config> {
+    pub owner: AccountIdOf<T>,
+    pub time: MomentOf<T>,
+    pub amount: Balance,
+    pub payment_reference: BoundedString<T::MaxRequestPaymentReferenceSize>,
+    pub details: Option<BoundedString<T::MaxRequestDetailsSize>>,
+    pub status: RequestStatus<T>,
+}
+
+impl<T: Config> Clone for DepositRequest<T> {
+    fn clone(&self) -> Self {
+        Self {
+            owner: self.owner.clone(),
+            time: self.time,
+            amount: self.amount,
+            payment_reference: self.payment_reference.clone(),
+            details: self.details.clone(),
+            status: self.status.clone(),
+        }
+    }
+}
+
+impl<T: Config> PartialEq for DepositRequest<T> {
+    fn eq(&self, other: &Self) -> bool {
+        self.owner == other.owner
+            && self.time == other.time
+            && self.amount == other.amount
+            && self.payment_reference == other.payment_reference
+            && self.details == other.details
+            && self.status == other.status
+    }
+}
+
+impl<T: Config> Eq for DepositRequest<T> {}
```

Comment:
It is necessary to implement these traits manually, because there are some problems with `derive` for `T::MaxRequestPaymentReferenceSize` & `T::MaxRequestDetailsSize`

Reply:
You can also use derivative crate to solve this. [Here](https://github.com/sora-xor/sora2-common/blob/f38ef9d584d6dfe341af5b2788c5446528a53bf4/pallets/types/src/types.rs#L333) is example